### ISSUE TITLE
 Drag and Drop columns with react virtualized

### ISSUE THREAD

**Issue by samichtioui** [clauderic/react-sortable-hoc#218]
I'm looking for an exmple of drag & drop columns with react-sortable-hoc and react-virtualized, I have not figured out how I can do it. Plz help me ...

```javascript
<Table
                                        height={height}
                                        width={1500}
                                        headerHeight={styles.rowHeight}
                                        rowHeight={styles.rowHeight}
                                        rowCount={tableData.length}
                                        rowGetter={getRow}
                                        headerStyle={styles.headerRow}
                                        rowStyle={getRowStyle}
                                        headerClassName="headerColumn"
                                    >
                                        {
                                            columns.map((line, key) => (
                                                <Column
                                                    key={key}
                                                    width={line.width}
                                                    dataKey={line.dataKey}
                                                    label={line.label}
                                                />
                                            ))
                                        }
                                    </Table>
```
@clauderic 
I succeeded for the rows but not for columns by following this link: https://embed.plnkr.co/N4iYHYE9gPLp4Y3NHbhk/

**Comment by samichtioui** [comment]
@chrisvasz can u help me pls?
I created an issue [stackoverflow](https://stackoverflow.com/questions/44454980/drag-and-drop-columns-with-react-sortable-hoc)

**Comment by lrecknagel** [comment]
I am also really looking for an idea how this can be achieved. I already tried to the extent the Column component but no luck with that. Its also interesting how to implement that with a MultiGrid example.

**Comment by szabolcsx** [comment]
Hi @samichtioui,

I have been struggling with the same thing but I managed to put together something which works.
You can find it here: https://jsfiddle.net/szabolcsx/ygysfogt/2/

Basically you need to provide a custom row renderer which must be wrapped in a SortableContainer HOC. Then you have to use custom header cell renderers which must be wrapped in a SortableElement HOC.

The code could use come cleanup but it will get you going.

I hope it helps.

**Comment by silvenon** [comment]
@szabolcsx awesome! ❤️  I based the following example off of yours:

```jsx
import React from 'react';
import { Table, Column, defaultTableHeaderRowRenderer, defaultTableHeaderRenderer, AutoSizer } from 'react-virtualized';
import { sortableContainer, sortableElement, arrayMove } from 'react-sortable-hoc';
import 'react-virtualized/styles.css';

const SortableHeaderRowRenderer = sortableContainer(defaultTableHeaderRowRenderer);
const SortableHeaderRenderer = sortableElement(props =>
  <div>{defaultTableHeaderRenderer(props)}</div>);

export default class extends React.Component {
  state = {
    data: [
      { col1: 'col1 data', col2: 'col2 data', col3: 'col3 data' },
      { col1: 'col1 data', col2: 'col2 data', col3: 'col3 data' },
      { col1: 'col1 data', col2: 'col2 data', col3: 'col3 data' },
    ],
    cols: [
      { dataKey: 'col1', label: 'Column 1' },
      { dataKey: 'col2', label: 'Column 2' },
      { dataKey: 'col3', label: 'Column 3' },
    ],
  }

  _rowGetter = ({ index }) => this.state.data[index];

  _handleSortEnd = ({ oldIndex, newIndex }) => {
    this.setState(state => ({
      cols: arrayMove(state.cols, oldIndex, newIndex),
    }));
  };

  _headerRowRenderer = props =>
    <SortableHeaderRowRenderer
      {...props}
      axis="x"
      helperClass="sortable-helper"
      lockAxis="x"
      style={{
        ...props.style,
        userSelect: 'none',
        cursor: 'default',
      }}
      onSortEnd={this._handleSortEnd}
    />;

  _headerRenderer = props =>
    <SortableHeaderRenderer
      {...props}
      index={props.columnData.index}
    />;

  render() {
    const { data, cols } = this.state;
    return (
      <div>
        <style>{`
          .sortable-helper {
            text-transform: uppercase;
            font-weight: 700;
          }
        `}</style>
        <Table
          headerHeight={20}
          headerRowRenderer={this._headerRowRenderer}
          height={110}
          rowCount={data.length}
          rowGetter={this._rowGetter}
          rowHeight={30}
          width={300}
        >
          {cols.map((col, index) =>
            <Column
              {...col}
              columnData={{ index }}
              headerRenderer={this._headerRenderer}
              key={col.dataKey}
              width={100}
            />
          )}
        </Table>
      </div>
    );
  }
}
```

The only thing that still bugs me is that the wider the columns are, the harder it is to sort them. I don't know how to decrease that threshold…

**Comment by silvenon** [comment]
It turns out that's because the headers aren't direct descendants of their sortable container. I don't know if this is a shortcoming of react-virtualized or react-sortable-hoc…

**Comment by silvenon** [comment]
I got it! I have no idea how I haven't seen this before, but `headerRowRenderer` itself allows us to make header cells direct descendants of the sortable container.

Complete example: https://codesandbox.io/s/vq8n2o68v5

For `index` I'm extracting the number from the column key. I'm not sure if that's good practice, if someone can figure out a way to read `columnData` from columns in `headerRowRenderer`, they could pass the index that way.

**Comment by clauderic** [comment]
This looks great @silvenon, thanks for taking the time to put an example together and sharing it 👍 

Would you be interested in submitting a PR to add this to the [examples folder](https://github.com/clauderic/react-sortable-hoc/tree/master/examples)? 😄

**Comment by szabolcsx** [comment]
@clauderic @silvenon I've met with the same issue as before. If you have a narrower column on the right, and you are trying to drag a wide column from left to right, you get a weird jumping anomaly.

Here is an example with the issue (it is based on @silvenon's code, I just added props for the column widths): https://codesandbox.io/s/1okw8py3y4

Just try to drag COL 1 over COL2, and then move a cursor a little to the left.
The issue is only present when you are trying to move a column from left to right.

I think the issue comes from the way SortableContainer would scroll the scrolling container, however in this case, the table header cannot be scrolled. If I increase the width of the header row, the jumping stops, however that messes up other things with react virtualized:
``` HTML
<style>
  .ReactVirtualized__Table__headerRow {
    width: 1200px !important;
  }
</style>
```

I tried implementing the same functionality with two grids and scroll sync. The same issue is present, so using a scrollable container doesn't solve the problem.
JSFiddle: https://jsfiddle.net/0mnwgqey/2/

Any ideas how to fix this?
